crates/sui-proc-macros/src/lib.rs

```diff
@@ -227,52 +230,217 @@ pub fn sim_test(args: TokenStream, item: TokenStream) -> TokenStream {
     result.into()
 }
 
-#[proc_macro_attribute]
-pub fn use_checked_arithmetic(_attr: TokenStream, item: TokenStream) -> TokenStream {
-    let input_item = parse_macro_input!(item as syn::Item);
+#[proc_macro]
+pub fn checked_arithmetic(input: TokenStream) -> TokenStream {
+    let input_file = CheckArithmetic.fold_file(parse_macro_input!(input));
+
+    let output_items = input_file.items;
+
+    let output = quote! {
+        #(#output_items)*
+    };
+
+    TokenStream::from(output)
+}
 
+#[proc_macro_attribute]
+pub fn with_checked_arithmetic(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    let input_item = parse_macro_input!(item as Item);
     match input_item {
-        syn::Item::Fn(input_fn) => {
+        Item::Fn(input_fn) => {
             let transformed_fn = CheckArithmetic.fold_item_fn(input_fn);
             TokenStream::from(quote! { #transformed_fn })
         }
-        syn::Item::Impl(input_impl) => {
+        Item::Impl(input_impl) => {
             let transformed_impl = CheckArithmetic.fold_item_impl(input_impl);
             TokenStream::from(quote! { #transformed_impl })
         }
-        _ => panic!(
-            "The use_checked_arithmetic attribute can only be applied to functions and impl blocks"
-        ),
+        _ => {
+            let err = syn::Error::new_spanned(
+                input_item,
+                "The with_checked_arithmetic attribute can only be applied to functions and impl blocks",
+            );
+            err.to_compile_error().into()
+        }
     }
 }
 
 struct CheckArithmetic;
 
+impl CheckArithmetic {
+    fn maybe_skip_macro(&self, attrs: &mut Vec<Attribute>) -> bool {
+        if let Some(idx) = attrs
+            .iter()
+            .position(|attr| attr.path().is_ident("skip_checked_arithmetic"))
+        {
+            // Skip processing macro because it is annotated with
+            // #[skip_checked_arithmetic]
+            attrs.remove(idx);
+            true
+        } else {
+            false
+        }
+    }
+
+    fn process_macro_contents(
+        &mut self,
+        tokens: proc_macro2::TokenStream,
+    ) -> syn::Result<proc_macro2::TokenStream> {
+        // Parse the macro's contents as a comma-separated list of expressions.
+        let parser = Punctuated::<Expr, Token![,]>::parse_terminated;
+        let Ok(exprs) = parser.parse(tokens.clone().into()) else {
+            return Err(syn::Error::new_spanned(
```

Comment:
Could we save this error for a case where we see an arithmetic operator in the token stream?

Reply:
We could, but I already spent more time on this than I intended to, and I didn't run into any cases of macros that needed to be excluded in any of the files that I converted, so I don't think its worth the effort for now.